test/Microsoft.ML.Functional.Tests/Explainability.cs

```diff
@@ -49,6 +51,43 @@ public void GlobalFeatureImportanceWithPermutationFeatureImportance()
                 Common.AssertMetricsStatistics(metricsStatistics);
         }
 
+        /// <summary>
+        /// GlobalFeatureImportance: PFI can be used to compute global feature importance. Here it is used with a model loaded from disk.
+        /// </summary>
+        [Fact]
+        public void GlobalFeatureImportanceWithPermutationFeatureImportanceWithLoadedModel()
+        {
+            var mlContext = new MLContext(seed: 1);
+
+            // Get the dataset
+            var data = mlContext.Data.LoadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Regression.Trainers.Sdca());
+
+            // Fit the pipeline and transform the data.
+            var model = pipeline.Fit(data);
+
+            var modelAndSchemaPath = GetOutputPath("TestFunctionalTestPFI.zip");
+            mlContext.Model.Save(model, data.Schema, modelAndSchemaPath);
+
+            ITransformer loadedModel;
+            using (var fs = File.OpenRead(modelAndSchemaPath))
+                loadedModel = mlContext.Model.Load(modelAndSchemaPath, out var schema);
+
+            var transformedData = loadedModel.Transform(data);
+            var linearPredictor = (loadedModel as TransformerChain<ITransformer>).LastTransformer as RegressionPredictionTransformer<LinearRegressionModelParameters>;
```

Comment:
>RegressionPredictionTransformer<LinearRegressionModelParameters> [](start = 101, length = 64)

Wouldn't this work if we write `RegressionPredictionTransformer<object>`?

Reply:
It doesn't work since the cast to `RegressionPredictionTransformer<object>` returns null, and having a null linearPredictor makes it impossible to use the PFI method.

I've just tested this, and this happens both with or without the changes I've made to the Prediction Transformers in this pull request.

Notice that originally the `.LastTransformer ` would return a `RegressionPredictionTransformer<IPredictorProducing<float>>` object, but with the changes in this pull request it now returns a `RegressionPredictionTransformer<LinearRegressionModelParameters>` object. So, as I've stated, in both cases a cast to `RegressionPredictionTransformer<object>` returns null.

Reply:
Oh, I see. Would it work with `ISingleFeaturePredictionTransformer<object>`? There is a test in the functional tests project that loads a model and extracts the predictor:
[https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Functional.Tests/ModelFiles.cs#L193](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Functional.Tests/ModelFiles.cs#L193)

Would something similar work here?

---
In reply to: [329679240](https://github.com/dotnet/machinelearning/pull/4262#discussion_r329679240) [](ancestors = 329679240)

Reply:
Yes, a cast to `ISingleFeaturePredictionTransformer<object>` works in this test.

I've tested this with the changes I've made in this pull request, and without them. And it works in both cases.

Reply:
So, should I change the cast in here to `ISingleFeaturePredictionTransformer<object>` for some reason?

Or are there other casts you want me to try out?

Reply:
If it works without the change, then why do we need this change at all?

Is it because we don't expect the user to know which version of `PermutationFeatureImportance` to call (regression/classification)?

---
In reply to: [330194364](https://github.com/dotnet/machinelearning/pull/4262#discussion_r330194364) [](ancestors = 330194364)

Reply:
> If it works without the change, then why do we need this change at all?

I don't believe users should be using `ISingleFeaturePredictionTransformer<object>` in a mainline scenario, like using PFI. They should be able to use the exact same type as what they got back when they created the model in the first place.

```C#
TransformerChain<RegressionPredictionTransformer<OlsModelParameters>> model = pipeline.Fit(data);
```

When I save and then load the above model, I should get back a `TransformerChain<RegressionPredictionTransformer<OlsModelParameters>>` object. But with the current code, the loaded object is a TransformerChain<RegressionPredictionTransformer<IPredictorProducing<float>>>

The root issue here is that the type of object that we get back when loading a model from disk is not the same as the type of object that is created in the first place. Thus, certain casts that work when you haven't saved the model to disk will work, and then fail for the same model after it is saved and then loaded.

Reply:
I hadn't tried to cast to `ISingleFeaturePredictionTransformer<object>` until you've mentioned it in here, so I didn't know PFI would work that way, even without my changes. So now I am unsure if the changes I've donde in this pull request are actually necessary.

There is still a problem that isn't present in the test cases, but that [arises in the sample case of PFI with Regression](https://github.com/antoniovs1029/machinelearning/blob/8e190e480ae2270da6ee0f587f1a9cb7a52976fe/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Regression/PermutationFeatureImportanceLoadFromDisk.cs#L79). Specifically what happens is that the code tries to acces `linearPredictor.Model.Weights[i]` and if `linearPredictor ` was created using the cast you are proposing:
`var linearPredictor =  (model as TransformerChain<ITransformer>).LastTransformer as ISingleFeaturePredictionTransformer<object>`

then linearPredictor doesn't have access to the Model.Weights member, and the compiler throws an error.

If I change the cast to `ISingleFeaturePredictionTransformer<OlsModelParameters>` then the compiler doesn't complain, and it works with the changes I've made in this pull request... but it doesn't work without them (the cast returns a null reference).

So, in short, it seems to me that if the user casts the last transformer to `ISingleFeaturePredictionTransformer<object>` then s/he will be able to use PFI even without my changes, but s/he might not be able to access the members of the parameters.

I don't know if there's a work around I am not seeing for this. But I will keep on trying other casts and see if I find one that works.

Reply:
> I don't believe users should be using ISingleFeaturePredictionTransformer<object> in a mainline scenario, like using PFI. They should be able to use the exact same type as what they got back when they created the model in the first place.

So I understand that the way to go is to keep my changes and not expect the user to use the  `ISingleFeaturePredictionTransformer<object>` cast, right?

Should I also add a test where loading a model from disk and casting it to  `ISingleFeaturePredictionTransformer<object>` doesn't work? Such as the situation I've described in my last comment in here?

Reply:
Agree on "They should be able to use the exact same type as what they got back when they created the model in the first place.". I think we should be expecting users to cast it to the same type they saved it not some generic type with object ... 

---
In reply to: [330212148](https://github.com/dotnet/machinelearning/pull/4262#discussion_r330212148) [](ancestors = 330212148)